File: blueprints/api/impostazioni.py
"""API REST per le preferenze dell'app (timer di default, contesto AI,
attrezzatura disponibile, altezza e target giornalieri).

La scelta del modello AI/Ollama vive invece in `blueprints/api/chat.py`: è
strettamente legata al catalogo provider (`services/ai.py`), non una
preferenza indipendente come queste.
"""

from flask import Blueprint, request

from models import Impostazione
from schemas import ApiError, api_ok
from services.ai_tools import (
    ErroreStrumento,
    imposta_attrezzatura,
    imposta_preferenza,
)

bp = Blueprint("api_impostazioni", __name__, url_prefix="/api")

# Preferenze numeriche gestite tutte allo stesso modo: il controllo dei limiti
# sta in `imposta_preferenza`, che è anche lo strumento usato dall'assistente.
# L'altezza serve al BMI nella pagina del peso e vale sempre; i target
# giornalieri disegnano le linee di riferimento nella sezione Salute.
CHIAVI_NUMERICHE = (
    "timer_default_sec",
    "analisi_n_sessioni",
    "altezza_cm",
    "target_kcal",
    "target_proteine_g",
    "target_carboidrati_g",
    "target_grassi_g",
    "target_sonno_minuti",
)


def _dati():
    dati = {chiave: Impostazione.get_int(chiave, 0) for chiave in CHIAVI_NUMERICHE}
    dati["attrezzatura_disponibile"] = (
        Impostazione.get("attrezzatura_disponibile") or ""
    )
    return dati


@bp.get("/impostazioni")
def leggi_impostazioni_route():
    return api_ok(_dati())


@bp.patch("/impostazioni")
def modifica_impostazioni_route():
    corpo = request.get_json(force=True, silent=True) or {}
    try:
        for chiave in CHIAVI_NUMERICHE:
            if chiave in corpo:
                imposta_preferenza(chiave, corpo[chiave])
        if "attrezzatura_disponibile" in corpo:
            imposta_attrezzatura(corpo["attrezzatura_disponibile"])
    except ErroreStrumento as exc:
        raise ApiError("VALIDATION_ERROR", str(exc), 422)
    return api_ok(_dati())

File: frontend/src/api/impostazioni.ts
import { api } from "@/lib/api"

/** Le preferenze numeriche valgono 0 quando non sono impostate: senza altezza
 *  non si mostra il BMI, senza target non si disegna la linea di riferimento. */
export interface Impostazioni {
  timer_default_sec: number
  analisi_n_sessioni: number
  altezza_cm: number
  target_kcal: number
  target_proteine_g: number
  target_carboidrati_g: number
  target_grassi_g: number
  target_sonno_minuti: number
  attrezzatura_disponibile: string
}

export type ModificaImpostazioni = Partial<Impostazioni>

export const impostazioniApi = {
  leggi: () => api.get<Impostazioni>("/impostazioni"),
  modifica: (dati: ModificaImpostazioni) => api.patch<Impostazioni>("/impostazioni", dati),
}
